packages/nitro-protocol/contracts/ninja-nitro/SingleChannelAdjudicator.sol

```diff
@@ -308,154 +316,117 @@ contract SingleChannelAdjudicator is
-        safeToDelete = true; // switched to false if there is an item remaining with amount > 0
-        uint256 surplus = initialHoldings; // tracks funds available during calculation
-        uint256 k = 0; // indexes the `indices` array
-
-        // copy allocation
-        for (uint256 i = 0; i < allocation.length; i++) {
-            newAllocation[i].destination = allocation[i].destination;
-            newAllocation[i].amount = allocation[i].amount;
+    /**
+     * @notice Transfers as many funds escrowed against `guarantorChannelId` as can be afforded for a specific destination in the beneficiaries of the __target__ of that channel. Checks against the storage in this contract.
+     */
+    function claim(
+        bytes32 guarantorChannelId,
+        bytes32 targetChannelId,
+        ChannelDataAlt calldata guaranteeCDL,
+        ChannelDataAlt calldata targetCDL,
+        uint256[][] memory indices
+    ) external {
+        // CHECKS
+        for (uint256 i = 0; i < indices.length; i++) {
+            _requireIncreasingIndices(indices[i]);
         }
-
-        // for each guarantee destination
-        for (uint256 j = 0; j < guarantee.destinations.length; j++) {
-            if (surplus == 0) break;
-            for (uint256 i = 0; i < newAllocation.length; i++) {
-                if (surplus == 0) break;
-                // search for it in the allocation
-                if (guarantee.destinations[j] == newAllocation[i].destination) {
-                    // if we find it, compute new amount
-                    uint256 affordsForDestination = min(allocation[i].amount, surplus);
-                    // decrease surplus by the current amount regardless of hitting a specified index
-                    surplus -= affordsForDestination;
-                    if ((indices.length == 0) || ((k < indices.length) && (indices[k] == i))) {
-                        // only if specified in supplied indices, or we if we are doing "all"
-                        // reduce the new allocationItem.amount
-                        newAllocation[i].amount -= affordsForDestination;
-                        // increase the relevant payout
-                        payouts[k] += affordsForDestination;
-                        totalPayouts += affordsForDestination;
-                        // move on to the next supplied index
-                        ++k;
-                    }
-                    break; // start again with the next guarantee destination
-                }
-            }
+        Outcome.OutcomeItem[] memory guarantorOutcome = abi.decode(
+            guaranteeCDL.outcomeBytes,
+            (Outcome.OutcomeItem[])
+        );
+        address payable guarantor = AdjudicatorFactory(adjudicatorFactoryAddress).getChannelAddress(
+            guarantorChannelId
+        );
+        address targetChannelAddress = AdjudicatorFactory(adjudicatorFactoryAddress)
+            .getChannelAddress(targetChannelId);
```

Comment:
I believe this is better interpreted as a type assertion, rather than a constructor call. It's there to tell the compiler to look up the abi of `AdjudicatorFactory`, so that (for example) the `.getChannelAddress` can be type checked. We aren't creating any new contracts here, we are just telling the compiler that there is a contract already deployed at `adjudicatorFactoryAddress`, and it has a certain interface. So I don't think there are any seriously adverse gas implications.

Of course we could make a new variable with the existing address, type assert it once and then reuse that. There might be a tiny difference in gas consumption.

Reply:
572a33e

Reply:
Digging into the test artifacts on circle, it seems that gas consumption increased by ~ 5 or 10 gas in some of the test cases 🤷‍♂️